lib/Conversion/CombToCGGI/CombToCGGI.cpp

```diff
@@ -151,23 +196,53 @@ struct ConvertTruthTableOp : public OpConversionPattern<TruthTableOp> {
   }
 };
 
+// ConvertSecretCastOp removes secret.cast operations between multi-bit secret
+// integers and tensors of single-bit secrets.
+struct ConvertSecretCastOp : public OpConversionPattern<secret::CastOp> {
+  ConvertSecretCastOp(mlir::MLIRContext *context)
+      : OpConversionPattern<secret::CastOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      secret::CastOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    // If this is a cast from secret<i8> to secret<tensor<8xi1>> or vice
+    // versa, replace with the cast's input.
+    auto inputTy =
+        cast<secret::SecretType>(op.getInput().getType()).getValueType();
+    auto outputTy =
+        cast<secret::SecretType>(op.getOutput().getType()).getValueType();
+
+    if (equivalentMultiBitAndTensor(inputTy, outputTy) ||
+        equivalentMultiBitAndTensor(outputTy, inputTy)) {
+      rewriter.replaceOp(op, adaptor.getInput());
+      return success();
+    }
+
+    return failure();
+  }
+};
+
 struct CombToCGGI : public impl::CombToCGGIBase<CombToCGGI> {
   void runOnOperation() override {
     MLIRContext *context = &getContext();
     auto *module = getOperation();
-    SecretTypeConverter typeConverter(context);
+
+    // TODO(https://github.com/google/heir/issues/250): The lutSize here is
+    // fixed on the assumption that the comb dialect is using ternary LUTs.
+    // Generalize lutSize by doing an analysis pass on the input combinational
+    // operations and integers.
```

Comment:
Optional: You might not need an analysis pass. I imagine you could just walk the IR here, type switch to match on comb ops, and use a max to produce either 2 or 3 based on whether there are any lut3's in the IR.

Reply:
Yes! I was considering walking this as well. For now, I'm going to keep this as hard-coded, since the only comb operation we have supported for Yosys to output is LUT 3's.

The next change here would be if we do builtin Yosys gates and have a backup bitwidth size of 1 in case there's not LUTs found.